로그인 기능 개발

Habist opened the pull request:
## 작업 내용
- 로그인 정보를 담을 `UserLogin` 모델을 추가하였습니다.
- 로그인 요청을 처리할 컨트롤러 매핑을 추가하였습니다.
- 로그인 요청을 처리할 비즈니스 로직을 추가하였습니다.
   - 로그인 성공 시 세션에 유저의 정보를 저장하도록 하였습니다.
- 로그인 정보의 아이디와 패스워드로 검색하는 쿼리를 추가하였습니다.

>  참조이슈
> [#7]

Habist committed:
feat: [#7] 로그인 기능 개발

로그인 정보를 담을 모델을 추가했습니다.
로그인 요청을 처리할 컨트롤러 매핑을 추가했습니다.
로그인 요청을 처리할 비즈니스 로직을 추가하였습니다.
- 로그인 성공 시 세션에 유저의 정보를 저장합니다.
로그인 정보의 아이디와 패스워드로 검색하는 쿼리를 추가했습니다.

f-lab-dev commented on src/main/java/com/campool/mapper/UserMapper.java:
```diff
@@ -14,4 +14,7 @@
     @Select("SELECT * FROM USER WHERE id = #{id}")
     UserSignUp findById(String id);
 
+    @Select("SELECT * FROM USER WHERE id = #{id} AND password = #{password}")
```
1. 쿼리를 작성할 땐 어떤 컬럼을 불러오는지 명확하게 표현해주는 것이 좋습니다~
2. 쿼리를 작성하는 방법은 이런 어노테이션 방식과 xml방식이 있는데 어떤 상황에 어떤 방식이 좋은지도 알아보시면 좋겠습니다~

Habist replied:
> 쿼리를 작성할 땐 어떤 컬럼을 불러오는지 명확하게 표현해주는 것이 좋습니다~
- 넵 불러오는 컬럼에 대한 명시를 추가하겠습니다!
> 쿼리를 작성하는 방법은 이런 어노테이션 방식과 xml방식이 있는데 어떤 상황에 어떤 방식이 좋은지도 알아보시면 좋겠습니다~
- 자바의 애노테이션 방식은 xml작성과 그에 따른 오버헤드 없이 간단하게 sql 매핑을 사용할 수 있는 장점이 있지만 xml기반의 방식보다는 표현성과 유연함에 한계가 있다는 점입니다. 복잡한 쿼리를 바탕으로 유연하고 다이나믹한 매핑이 필요하면 xml을 사용하는 게 좋고 간단한 매핑에 대해서는 애노테이션 방식도 괜찮은 선택이라고 느껴집니다.
이 부분은 프로젝트를 계속 진행하면서 필요한 상황이 오면 도입을 고려해도 괜찮을까요??

f-lab-dev replied:
네 복잡한 쿼리를 작성하는데에 있어 어노테이션 방식은 가독성이 좋지 않습니다~
방식을 하나로 통일하는것도 고민해보셔도 될 것 같아요

f-lab-dev commented on src/main/java/com/campool/service/UserService.java:
```diff
@@ -30,4 +33,14 @@ public boolean isDuplicate(String id) {
+        UserSignUp userSignUp = userMapper
+                .findByIdAndPassword(userLogin.getId(), encryptor.encrypt(userLogin.getPassword()));
+        if (userSignUp == null) {
+            throw new RuntimeException(new NotFoundException("로그인에 실패했습니다."));
```
여기서는 굳이 감쌀 필요 없이 `NotFoundException`만 던져줘도 될 것 같은데 감싸주신 이유가 있으실까요?

Habist replied:
`NotFoundException`은 checked exception이라 login을 호출한 쪽에서 반드시 예외처리를 강제한다는 점이 있습니다.
 로그인 로직에서 발생시킨 예외는 로그인을 실패했다고 알리는 용도라고 생각해 예외에 대한 처리를  login 메소드를 호출한 쪽에서 처리할 필요가 없다고 판단했습니다!

f-lab-dev replied:
그럼 `NotFoundException` 비슷한거를 `RuntimeException`을 상속받아 만들어줘도 될 것 같네요~

Habist replied:
`UserService`의 상황에 맞는 예외를 추가하였습니다!

f-lab-dev commented on src/main/java/com/campool/service/UserService.java:
```diff
@@ -30,4 +33,14 @@ public boolean isDuplicate(String id) {
         return userMapper.findById(id) != null;
     }
 
+    public void login(UserLogin userLogin, HttpSession session) {
```
1. `UserService`는 어떤 책임을 갖고 있는 클래스인가요?
2. 로그인은 그 책임에 맞는 동작일까요?

Habist replied:
> UserService는 어떤 책임을 갖고 있는 클래스인가요?
- 사용자 정보를 기반으로 하는 요청이 들어왔을 때 요청 데이터를 **어떻게 조작**해서 그 기능을 **구현**할 것인지를 책임지고 있는 클래스입니다.
> 로그인은 그 책임에 맞는 동작일까요?
-  로그인은 사용자의 정보를 받아서 인증을 위해 어떻게 처리할 것인지를 정의를 해야하는데 Controller는 사용자 요청 정보의 유효성 검증과 어떤 비즈니스 로직을 선택해 처리할 것인지를 책임지고 Service는 요청 정보를 어떻게 조작하여 원하는 기능을 구현할 것인지를 책임진다고 생각했습니다. 그래서 로그인은 **"사용자 정보를 기반으로 해당하는 정보가 DB에 존재하면 세션에 사용자 정보를 등록한다"** 라는 의미로 요청 정보를 **어떻게 조작**해서 로그인 기능을 **구현**할 것이지를 담당하는 Service 클래스가 적당하다 생각하였습니다.

f-lab-dev replied:
네 Service 레이어가 적당하지만 **인증**이 `UserService`의 책임인지는 고민을 해봐야할 것 같습니다~

또한 세션 방식의 로그인 외에 토큰 방식의 로그인이 추가된다고 가정했을 때 이 클래스는 변경에 유연한가도 고민해봐야할 포인트일거고요~

Habist replied:
사용자 정보를 **확인**하는 부분은 `UserService`에서 처리하도록 하였고, 그 결과를 통해 **인증**하는 부분은 인증을 담당하는 서비스를 만들어서 그 클래스가 책임을 담당하도록 하였습니다.
또 로그인 방식이 바뀌는 경우를 고려해 인터페이스를 의존하게 만들어 구체적인 방식이 변경되더라도 다른 코드는 변경될 필요가 없도록 수정했습니다!

f-lab-dev commented on src/main/java/com/campool/service/UserService.java:
```diff
@@ -30,4 +33,14 @@ public boolean isDuplicate(String id) {
+        if (userSignUp == null) {
+            throw new RuntimeException(new NotFoundException("로그인에 실패했습니다."));
+        } else {
+            session.setAttribute("userSignUp", userSignUp);
```
1. `"userSignUp"` 같은 세션의 키로 사용되는 것은 상수로 빼두는게 유지보수성에 좋습니다~
2. 이런 동작은 캡슐화시켜서 의도에 맞게 사용할 수 있도록 해주는게 좋습니다

Habist replied:
- 세션을 설정하는 책임을 가진 클래스로 분리하였고 그 클래스 안에 상수를 정의했습니다. 
- 인증된 사용자인지 판단하는 부분을 좀 더 확실한 의미를 가지도록 메소드로 분리하였습니다.

Habist committed:
improve: [#7 - 리뷰반영] 쿼리문 수정

쿼리문에서 어떤 컬럼을 가져올지 명확하게 명시하도록 수정하였습니다.

Habist committed:
improve: [#7 - 리뷰반영] 사용자 인증 로직 분리

사용자를 인증하는 로직을 더 명확한 의미를 갖도록 메소드로 분리하였습니다.
세션을 다루는 책임을 가지는 클래스를 만들었습니다.

f-lab-dev commented on src/main/java/com/campool/auth/SessionAuth.java:
```diff
@@ -0,0 +1,14 @@
+import com.campool.model.UserSignUp;
+import javax.servlet.http.HttpSession;
+
+public class SessionAuth {
```
코멘트 달아드린 것 기반으로 이 클래스의 존재에 대해서도 고민해보시면 좋겠네요~

> 네 Service 레이어가 적당하지만 인증이 UserService의 책임인지는 고민을 해봐야할 것 같습니다~
> 또한 세션 방식의 로그인 외에 토큰 방식의 로그인이 추가된다고 가정했을 때 이 클래스는 변경에 유연한가도 고민해봐야할 포인트일거고요~

Habist committed:
improve: [#7 - 리뷰반영] 사용자 정보 예외 클래스 추가

사용자 정보를 찾을 수 없는 경우에 사용되는 런타임 예외 클래스를 추가했습니다.

Habist committed:
improve: [#7 - 리뷰반영] 로그인 인증 서비스 추가

사용자 정보를 받아 처리하는 부분과 로그인 인증하는 부분을 분리하였습니다.
행위에 좀 더 적합한 이름으로 메스드 이름을 변경하였습니다.
인증 방식이 변경될 가능성을 고려해 인증 방식을 추상화한 인터페이스와
그것을 구현한 세션 인증 클래스를 추가하였습니다.

f-lab-dev commented on src/main/java/com/campool/service/SessionAuthService.java:
```diff
@@ -0,0 +1,26 @@
+
+@RequiredArgsConstructor
+@Service
+public class SessionAuthService implements AuthService {
```
👍

f-lab-dev commented on src/main/java/com/campool/model/UserLogin.java:
```diff
@@ -0,0 +1,22 @@
+@Getter
+@RequiredArgsConstructor
+@ToString
+public class UserLogin {
```
`UserLogin`은 이 클래스의 역할이 명확하지 않으므로 `UserLoginRequest`로 지어주는게 나을 것 같습니다~

Habist replied:
로그인 요청이니 요청의 의미를 담아주는 게 확실히 더 좋겠네요!

f-lab-dev commented on src/main/java/com/campool/service/UserService.java:
```diff
@@ -30,4 +32,18 @@ public boolean isDuplicate(String id) {
         return userMapper.findById(id) != null;
     }
 
+    public UserSignUp getByUserLogin(UserLogin userLogin) {
```
이 메소드는 로그인 외에도 더 널리 사용될 수 있을 것 같습니다.
`getByIdAndPw` 정도로 지어주면 더 재사용성이 높은 이름이 되지 않을까요?

Habist replied:
그러면 넘겨주는 매개변수도 Id와 Pw를 넘겨주도록 명시하는 게 좋을까요?

f-lab-dev replied:
네 저는 그게 나을 것 같습니다~

f-lab-dev commented on src/main/java/com/campool/controller/UserController.java:
```diff
@@ -15,9 +17,18 @@
 
+    @PostMapping("/users/login")
+    public void loginUser(@Valid UserLogin userLogin) {
+        UserSignUp userSignUp = userService.getByUserLogin(userLogin);
```
아래의 로직은 어떤 클래스가 가져야할 책임일까요?

1. ID와 비밀번호를 이용해 유저의 정보를 조회한다.
2. 정보가 맞는다면 인증처리를 한다.

Habist replied:
> ID와 비밀번호를 이용해 유저의 정보를 조회한다.
> 정보가 맞는다면 인증처리를 한다.

해당 로직들은 **사용자 데이터**인 ID와 PW를 **기반**으로 이루어지는 행위라 UserController가 책임지는 게 맞다고 생각했습니다.
책임의 중점을 요청 값을 기준으로 생각했는데 행위나 목표를 중점으로 생각해서 인증을 위한 AuthController가 책임지게 하는 것이 좋은 것인가요?

Habist replied:
> ID와 비밀번호를 이용해 유저의 정보를 조회한다.
> 정보가 맞는다면 인증처리를 한다.

해당 로직은 **사용자 데이터**인 ID와 PW를 **기반**으로 이루어지는 행위라 `UserController`가 책임지는 게 맞다고 생각했습니다.
**책임의 중점**을 **요청 값**을 기준으로 생각했는데 **행위**나 **목표**를 **중점**으로 생각해서 인증을 위한 컨트롤러가 책임지게 하는 것이 좋은 것인가요?
아니면 다른 부분을 더 생각해봐야 하는 것일까요??

f-lab-dev replied:
이 로직은 비지니스 로직입니다. 그러면 Service 레이어에 두는게 나을 것 같은데 `SessionAuthService`에 두는게 맞지 않을까요?

1. ID와 비밀번호를 이용해 유저의 정보를 조회한다.
2. 정보가 맞는다면 세션에 인증처리를 한다

Habist replied:
ID와 비밀번호를 이용해 유저 정보를 조회하는 것의 목표를 인증이라 생각하면 AuthService가 괜찮다고 생각듭니다.
메소드 시그니처를 명확한 명칭으로 변경하였고, 해당하는 로직은 인증 Service로 변경시켰습니다!

Habist committed:
refactor: [#7 - 리뷰반영] 로그인 요청 모델 이름 변경

로그인 요청을 확실하게 표현할 수 있도록 모델 이름을 변경하였습니다.

Habist committed:
refactor: [#7 - 리뷰반영] 인증 담당 메소드 서비스 변경

로그인 요청에 대한 인증을 담당하는 메소드를 인증 서비스에서 책임지도록
변경하였습니다.
로그인 요청을 처리하는 메소드의 시그니처를 좀 더 구체적으로 변경하였습니다.

f-lab-dev commented on src/main/java/com/campool/controller/UserController.java:
```diff
@@ -27,7 +27,8 @@ public void signUpUser(@Valid UserSignUp userSignUp) {
     public void loginUser(@Valid UserLoginRequest userLoginRequest) {
-        UserSignUp userSignUp = userService.getByUserLogin(userLoginRequest);
+        UserSignUp userSignUp = authService
+                .getByIdAndPw(userLoginRequest.getId(), userLoginRequest.getPassword());
```
이 로직은 `authService.authenticate()`에서 해도 될 것 같은데 Controller에서 해야할 이유가 있으실까요?

Habist replied:
service 레이어 끼리의 참조는 순환 참조를 일으킬 수 있다고 생각하고 있어서 무조건 컨트롤러에서 하는 쪽이 좋다고 생각했었네요..
service레이어를 단방향에서 참조하는 것은 문제가 되지 않을 거라 판단해서 `getByIdAndPw`는 `UserSerivice`에서 책임지도록 변경하였고 `SessionAuthService`에서 호출하도록 변경하였습니다!

f-lab-dev commented on src/main/java/com/campool/service/SessionAuthService.java:
```diff
@@ -11,8 +15,25 @@
     private final HttpSession session;
 
+    @Override
+    public UserSignUp getByIdAndPw(String id, String password) {
```
`유저의 정보를 조회해서 유효하다면 인증처리를 한다`는 동작은 `authenticate` 라는 범위에 들어있습니다

즉 합쳐도 될 것 같네요~

Habist replied:
넵 이 부분 리뷰 반영했습니다!

Habist committed:
refactor: [#7 - 리뷰반영] 사용자 정보 호출 메소드 책임 변경

사용자의 로그인 요청 값으로 사용자 정보를 가져오는 책임을 UserService에서 담당하도록
변경하였습니다.

UserController에서 담당하는 사용자 정보 처리 부분을 SessionAuthService에서
UserService를 통해 가져오도록 변경하였습니다.

f-lab-dev commented on src/main/java/com/campool/service/UserService.java:
```diff
@@ -30,4 +31,17 @@ public boolean isDuplicate(String id) {
+    }
+
+    private boolean isValidUser(UserSignUp userSignUp) {
+        return userSignUp != null;
```
보통 회원 탈퇴를 하면 바로 DB에서 삭제하는게 아니라 일정 기간동안은 탈퇴 여부를 마킹하고 남겨두는데 이 메소드를 통해 해당 기능을 검증할 수도 있겠네요~
